ledger/streaming/client/src/streaming_futures.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Custom Future implementations for the block streaming client
+
+use std::pin::Pin;
+use std::sync::{Arc, RwLock, TryLockError};
+
+use futures::Future;
+use futures::task::{Context, Poll};
+
+use mc_ledger_db::Ledger;
+use mc_transaction_core::{BlockData, BlockIndex};
+
+use crate::error::Error as ClientError;
+
+/// Future for writing blocks to LedgerDB using
+pub struct WriteBlock<'a, L: Ledger + 'a> {
+    ledger: Arc<RwLock<&'a mut L>>,
+    block_data: BlockData,
+}
+
+impl<'a, L: Ledger + 'a> WriteBlock<'a, L> {
+    /// Initialize future with ledger copy & block data
+    pub fn new(ledger: Arc<RwLock<&'a mut L>>, block_data: BlockData) -> Self {
+        Self {
+            ledger,
+            block_data,
+        }
+    }
+
+    /// TODO: Add more diverse error handling, some errors should cause the future to error
+    /// Attempt to write block, an error result here results in the future returning Poll:Pending
+    fn write_block(&self) -> Result<BlockIndex, ClientError> {
+        let block_index: BlockIndex;
+        match self.ledger.try_read() {
+            Ok(ledger_lock) => {
+                block_index = self.block_data.block().index;
+                if ledger_lock.num_blocks()? + 1 < block_index {
+                    return Err(ClientError::BlockIndexTooFar(block_index));
+                }
+            },
+            Err(err) => {
+                match err {
+                    TryLockError::WouldBlock => return Err(ClientError::Locked),
+                    _ => return Err(ClientError::LockPoisoned),
+                }
+            }
+        }
+
+        match self.ledger.try_write(){
+            Ok(mut ledger_lock) => {
+                ledger_lock.append_block(
+                    self.block_data.block(),
+                    self.block_data.contents(),
+                    self.block_data.signature().clone(),
+                )?;
+                Ok(block_index)
+            },
+            Err(err) => {
+                match err {
+                    TryLockError::WouldBlock => Err(ClientError::Locked),
+                    _ => Err(ClientError::LockPoisoned),
+                }
+            }
+        }
+    }
+}
+
+impl<'a, L: Ledger + 'a> Future for WriteBlock<'a, L> {
+    type Output = BlockIndex;
+
+    ///Attempt to write block and return the block index written
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.write_block() {
+            Ok(block_index) => {
+                Poll::Ready(block_index)
+            }
+            _ => {
+                cx.waker().wake_by_ref();
```

Comment:
Waking the task then returning Pending doesn't seem right. It seems easier and safer to reuse the StreamExt/FutureExt helpers than trying to implement `Future` ourselves.